### PR TITLE
v18/resource/cloudformation: keep finalizers when stacks still exist

### PR DESCRIPTION
**Pull request by kopiczko**
Fixes https://github.com/giantswarm/giantswarm/issues/4558.

### CODE REVIEW

File: service/controller/v18/resource/cloudformation/current.go [hunk 1]
```diff
@@ -37,7 +69,7 @@ func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interf
 		r.logger.LogCtx(ctx, "level", "debug", "message", "finding the guest cluster main stack outputs in the AWS API")
```
Comment (kopiczko) [inline service/controller/v18/resource/cloudformation/current.go]: Below only renaming.

File: service/controller/v18/resource/cloudformation/current.go [hunk 2]
```diff
@@ -23,11 +23,43 @@ func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interf
 
 	stackName := key.MainGuestStackName(customObject)
 
-	sc, err := controllercontext.FromContext(ctx)
+	ctlCtx, err := controllercontext.FromContext(ctx)
```
Comment (kopiczko) [inline service/controller/v18/resource/cloudformation/current.go]: Boyscout.

File: service/controller/v18/resource/cloudformation/current.go [hunk 3]
```diff
@@ -23,11 +23,43 @@ func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interf
...
+			in := &cloudformation.DescribeStacksInput{
+				StackName: aws.String(stackName),
+			}
+
+			out, err := ctlCtx.AWSClient.CloudFormation.DescribeStacks(in)
+			if cloudformationservice.IsStackNotFound(err) {
+				// This handling is far from perfect. We use different
+				// packages here. This is all going to be addressed in
+				// scope of
+				// https://github.com/giantswarm/giantswarm/issues/3783.
```
Comment (xh3b4sd) [inline service/controller/v18/resource/cloudformation/current.go]: We have the handling of stack statuses in several places and should maybe have some common library and understanding of how to work with these states. Like having a state machine with hooks or something. Is the issue referenced here where we want to address this or is this a different issue?

Comment (kopiczko) [reply service/controller/v18/resource/cloudformation/current.go]: I'll address it here. I even have some prototypes written.

File: service/controller/v18/resource/cloudformation/current.go [hunk 4]
```diff
@@ -23,11 +23,43 @@ func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interf
...
+		}
+
+		for _, stackName := range stackNames {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding stack %#q in the AWS API", stackName))
+
+			in := &cloudformation.DescribeStacksInput{
+				StackName: aws.String(stackName),
+			}
+
+			out, err := ctlCtx.AWSClient.CloudFormation.DescribeStacks(in)
```
[the remark this answers is not in the document]
Comment (kopiczko) [reply service/controller/v18/resource/cloudformation/current.go]:
```suggestion
			_, err := ctlCtx.AWSClient.CloudFormation.DescribeStacks(in)
```

[the remark this answers is not in the document]
Comment (kopiczko) [reply service/controller/v18/resource/cloudformation/current.go]: I can create suggestion for myself!

**Comment by kopiczko** [comment]
Because of the weirdness addressed here https://github.com/giantswarm/e2e-harness/pull/161 `v18` test were ran as `Cur` instead of `WIP` here https://circleci.com/workflow-run/64450a3e-0ce7-43fe-a3d9-1235dfb24f8f?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-workflow-link. All `Cur` tests are green. Going with that.

```
D 11/06 07:30:56 default/awsconfigs/ci-cur-f8ee3-6b509 cloudformationv18.GetCurrentState finding stack `cluster-ci-cur-f8ee3-6b509-guest-main` in the AWS API | aws-operator/service/controller/v18/resource/cloudformation/current.go:39 | event=delete | version=1161
D 11/06 07:30:56 default/awsconfigs/ci-cur-f8ee3-6b509 cloudformationv18.GetCurrentState found stack `cluster-ci-cur-f8ee3-6b509-guest-main` in the AWS API | aws-operator/service/controller/v18/resource/cloudformation/current.go:56 | event=delete | version=1161
D 11/06 07:30:56 default/awsconfigs/ci-cur-f8ee3-6b509 cloudformationv18.GetCurrentState keeping finalizer | aws-operator/service/controller/v18/resource/cloudformation/current.go:57 | event=delete | version=1161
D 11/06 07:30:56 default/awsconfigs/ci-cur-f8ee3-6b509 cloudformationv18.GetCurrentState finding stack `cluster-ci-cur-f8ee3-6b509-host-setup` in the AWS API | aws-operator/service/controller/v18/resource/cloudformation/current.go:39 | event=delete | version=1161
D 11/06 07:30:57 default/awsconfigs/ci-cur-f8ee3-6b509 cloudformationv18.GetCurrentState did not find stack `cluster-ci-cur-f8ee3-6b509-host-setup` in the AWS API | aws-operator/service/controller/v18/resource/cloudformation/current.go:52 | event=delete | version=1161
D 11/06 07:30:57 default/awsconfigs/ci-cur-f8ee3-6b509 cloudformationv18.GetCurrentState finding stack `cluster-ci-cur-f8ee3-6b509-host-main` in the AWS API | aws-operator/service/controller/v18/resource/cloudformation/current.go:39 | event=delete | version=1161
D 11/06 07:30:57 default/awsconfigs/ci-cur-f8ee3-6b509 cloudformationv18.GetCurrentState did not find stack `cluster-ci-cur-f8ee3-6b509-host-main` in the AWS API | aws-operator/service/controller/v18/resource/cloudformation/current.go:52 | event=delete | version=1161
```

**Pull request merged**